File: jot/Sources/BirtaJotCore/HostPrompt.swift
import Foundation

/// One step of a flow the page is driving, as it arrives over the bridge.
///
/// A literal reading of `HostPromptStep` in `shared/hostPrompt.ts`, and the
/// only thing about that seam this side owns: the questions, their order,
/// their validation rule and the composed report are all the page's, so a
/// second flow moving onto the seam adds a definition there and no Swift here.
/// That is the whole reason the seam exists, and it is what tells this apart
/// from `AgentRequest.swift` and `AgentReference.swift`, which are ports of
/// behaviour and carry headers naming where the two copies have diverged.
///
/// Parsing is strict, and a step that does not parse is not a step. A sheet
/// drawn from half a request would ask a question the page did not write, and
/// the page would then record the answer as if it had.
public enum HostPromptStep: Equatable {
    /// A free-text question.
    ///
    /// `required` and `maxLength` carry the message shown when they refuse,
    /// rather than a code this side would have to spell a sentence for: the
    /// wording is the page's, so both surfaces refuse a too-long title with
    /// the same words.
    case input(title: String, prompt: String, placeholder: String?,
               required: String?, maxLength: MaxLength?)
    /// A choice between named rows.
    case pick(title: String, placeholder: String?, rows: [Row])

    /// A ceiling, with the sentence shown when it refuses.
    public struct MaxLength: Equatable {
        public let value: Int
        public let message: String

        public init(value: Int, message: String) {
            self.value = value
            self.message = message
        }
    }

    public struct Row: Equatable {
        /// Returned verbatim as the answer when this row is chosen.
        public let id: String
        public let label: String
        /// The second line: what this row costs, or what it means.
        public let detail: String?

        public init(id: String, label: String, detail: String?) {
            self.id = id
            self.label = label
            self.detail = detail
        }
    }

    /// The title the sheet shows, whichever kind this is.
    public var title: String {
        switch self {
        case let .input(title, _, _, _, _): return title
        case let .pick(title, _, _): return title
        }
    }

    /// Read one step out of the message's `step` object.
    ///
    /// Returns nil for a kind this build does not draw, which the caller
    /// answers as `unsupported` rather than by staying silent: an unhandled
    /// message is dropped here with only a trace line (MAR-390), so silence
    /// would make a step this app cannot draw look exactly like one it
    /// correctly ignored.
    public static func parse(_ value: Any?) -> HostPromptStep? {
        guard let dict = value as? [String: Any], let kind = dict["kind"] as? String,
              let title = dict["title"] as? String else { return nil }
        let placeholder = dict["placeholder"] as? String

        switch kind {
        case "input":
            guard let prompt = dict["prompt"] as? String else { return nil }
            let required = (dict["required"] as? [String: Any])?["message"] as? String
            var maxLength: MaxLength?
            if let limit = dict["maxLength"] as? [String: Any],
               let count = (limit["value"] as? NSNumber)?.intValue,
               let message = limit["message"] as? String {
                maxLength = MaxLength(value: count, message: message)
            }
            return .input(title: title, prompt: prompt, placeholder: placeholder,
                          required: required, maxLength: maxLength)
        case "pick":
            // A pick with no rows is a question with no answers, which would
            // draw a sheet the user can only cancel.
            guard let raw = dict["rows"] as? [Any], !raw.isEmpty else { return nil }
            var rows: [Row] = []
            for entry in raw {
                guard let row = entry as? [String: Any],
                      let id = row["id"] as? String,
                      let label = row["label"] as? String else { return nil }
                rows.append(Row(id: id, label: label, detail: row["detail"] as? String))
            }
            return .pick(title: title, placeholder: placeholder, rows: rows)
        default:
            return nil
        }
    }

    /// The message to show for `value`, or nil when it is acceptable.
    ///
    /// A literal reading of `validateHostPromptInput`, including the trim:
    /// both surfaces have to agree on what an empty answer is, or one of them
    /// accepts a title made of spaces.
    public static func validate(_ value: String, required: String?,
                                maxLength: MaxLength?) -> String? {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        if let required, trimmed.isEmpty { return required }
        if let maxLength, trimmed.count > maxLength.value { return maxLength.message }
        return nil
    }
}

/// What Birta Writer for Mac reports about itself when a feedback report asks.
///
/// It names THIS host, which is why the page cannot gather it: the extension
/// reports its own version, VS Code's and the `birta.*` settings that differ
/// from their defaults, and none of those three exist here. The app reports
/// the app, macOS, and its own settings.
///
/// Never the note, its path, or the folder it is in. `shared/feedback/
/// compose.ts` is never given any of them, and this is the only thing that
/// reaches it from this side.
public struct HostDiagnostics: Equatable {
    public let appVersion: String
    public let systemVersion: String
    public let platform: String
    public let changedSettings: [String]

    public init(appVersion: String, systemVersion: String, platform: String,
                changedSettings: [String]) {
        self.appVersion = appVersion
        self.systemVersion = systemVersion
        self.platform = platform
        self.changedSettings = changedSettings
    }

    /// The `Diagnostics` shape `shared/feedback/compose.ts` reads.
    ///
    /// `extensionVersion` and `hostVersion` are the field names the composer
    /// prints as "Birta" and "Host"; what goes in them here is the app and
    /// macOS, so a report from this surface names what a reader would need to
    /// reproduce it rather than a VS Code that was never running.
    public var jsonObject: [String: Any] {
        [
            "extensionVersion": appVersion,
            "hostVersion": systemVersion,
            "platform": platform,
            "changedSettings": changedSettings,
        ]
    }
}

File: shared/hostPrompt.ts
/**
 * shared/hostPrompt.ts
 *
 * THE seam every interactive multi-step command flow goes through: a flow is
 * DEFINED once here, in TypeScript, and RENDERED by whichever host is running
 * (MAR-395).
 *
 * The problem it exists for. Every interactive flow the extension has is
 * written against `vscode.window.showInputBox` / `showQuickPick`, which is a
 * VS Code API and nothing else. A surface that is not VS Code therefore cannot
 * run any of them, and the alternative to a seam is porting each one to Swift
 * by hand: a second copy of the questions, their order, their validation and
 * their cancel semantics, drifting from the first as either side changes.
 * `AgentRequest.swift` and `AgentReference.swift` are the two such ports that
 * already exist, and their headers exist to name where the two have diverged.
 *
 * What crosses the wire is the STEP, one at a time, never the flow. Batching
 * would be fewer round trips and would lose the two things the flows actually
 * rely on: validation that runs as the user types, and the distinction between
 * cancelling a step and submitting it empty. Both are per-step facts.
 *
 * Validation is DECLARED rather than passed, because a function does not
 * survive `postMessage`. Each renderer turns the declaration back into whatever
 * its platform's live validation is, and `validateHostPromptInput` below is the
 * one implementation of the rule, so the two cannot disagree about what is
 * valid or about what the user is told.
 *
 * Dependency-free (no vscode, no DOM) so the extension, the webview and the
 * tests all import it.
 */

/** A free-text question. */
export interface HostPromptInputStep {
    readonly kind: "input";
    /** Window/sheet title. Carries the step's position, e.g. "(1 of 4)". */
    readonly title: string;
    /** The question itself. */
    readonly prompt: string;
    /** Greyed example text inside the field. Never a default value. */
    readonly placeholder?: string;
    /**
     * Refuse an empty answer, with the message the user is shown. Absent means
     * the step is optional, and an empty submission is a real answer rather
     * than a cancel.
     */
    readonly required?: { readonly message: string };
    /** Refuse an answer longer than `value`, with the message shown. */
    readonly maxLength?: { readonly value: number; readonly message: string };
}

/** One row of a `pick` step. */
export interface HostPromptRow {
    /** Returned verbatim as the step's answer when this row is chosen. */
    readonly id: string;
    readonly label: string;
    /** The second line: what this row costs, or what it means. */
    readonly detail?: string;
    /**
     * A VS Code codicon name (`github`, `mail`, `clippy`), drawn before the
     * label by the palette renderer and ignored by every other host. It is a
     * separate field rather than a `$(github) ` prefix inside `label` because
     * a host that is not VS Code would draw that prefix as the literal text.
     */
    readonly icon?: string;
}

/** A choice between named rows. */
export interface HostPromptPickStep {
    readonly kind: "pick";
    readonly title: string;
    /** The question. Drawn as the filter field's placeholder in the palette. */
    readonly placeholder?: string;
    readonly rows: readonly HostPromptRow[];
}

export type HostPromptStep = HostPromptInputStep | HostPromptPickStep;

/** The kinds a host has to be able to draw to serve any flow at all. */
export const ALL_HOST_PROMPT_KINDS: readonly HostPromptStep["kind"][] = ["input", "pick"];

/**
 * The message to show for `value`, or undefined when it is acceptable.
 *
 * Both renderers call this, so "what is valid" and "what the user is told"
 * have one home. A `pick` step needs none: its answers are its own rows.
 */
export function validateHostPromptInput(
    step: HostPromptInputStep,
    value: string,
): string | undefined {
    const trimmed = value.trim();
    if (step.required && !trimmed) return step.required.message;
    if (step.maxLength && trimmed.length > step.maxLength.value) return step.maxLength.message;
    return undefined;
}

/**
 * How a renderer answers one step: the text typed, the id of the row chosen,
 * or **null for a cancel**.
 *
 * Null and the empty string are different answers and the difference is
 * load-bearing: on an optional step, Escape abandons the flow and an empty
 * submission is a deliberate "nothing to add" that continues. VS Code's
 * `showInputBox` already draws exactly that distinction (`undefined` against
 * `""`), which is where it comes from.
 */
export type HostPromptAsk = (step: HostPromptStep, index: number) => Promise<string | null>;

/**
 * Whether `answer` is one this step could have produced.
 *
 * A renderer is supposed to refuse an invalid answer before it hands one back,
 * and both shipped renderers do: VS Code validates as the user types, and the
 * sheet puts the question again. This is the driver's own check, and it is
 * what stops a renderer's mistake becoming a report: a `pick` answer that
 * names no row the step offered was never chosen by anybody, and an `input`
 * that its own declared rules refuse is not an answer either.
 */
export function isAcceptableAnswer(step: HostPromptStep, answer: string): boolean {
    return step.kind === "pick"
        ? step.rows.some((row) => row.id === answer)
        : validateHostPromptInput(step, answer) === undefined;
}

/**
 * Put each step to `ask` in order, and return the answers.
 *
 * Returns null the moment any step is cancelled, and asks nothing after it:
 * the caller then has no partial answers to decide what to do with, which is
 * what makes "cancelling at any step leaves everything untouched" a property
 * of the driver rather than a rule each flow has to remember.
 *
 * An unacceptable answer STOPS the flow rather than being asked again. The
 * driver cannot usefully re-ask, because re-asking is the renderer's own job
 * and a renderer that has already broken its contract is the one thing here
 * that cannot be trusted to do it. Stopping asks nothing further and composes
 * nothing, which is the same outcome as a cancel and the safe reading of a
 * host that is not doing what it said.
 */
export async function runPromptFlow(
    steps: readonly HostPromptStep[],
    ask: HostPromptAsk,
): Promise<string[] | null> {
    const answers: string[] = [];
    for (let i = 0; i < steps.length; i++) {
        const answer = await ask(steps[i], i);
        if (answer === null || !isAcceptableAnswer(steps[i], answer)) return null;
        answers.push(answer);
    }
    return answers;
}
